Additional func

sharuiY opened the pull request:
Generate testcases for func startswith, endswith, and len.
Small change with or-logic.

sharuiY committed:
Generate seeds based on start end with

sharuiY committed:
Generate testcases for len func & change or logic

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -180,29 +184,138 @@ def _handle_compare_operation_cases(
+                            test_case[param_name] += "a" * (target_len - old_len)
+                        elif op == "<" and old_len >= target_len:
+                            test_case[param_name] = old_value[
+                                : target_len - 1 if target_len > 0 else 0
```
I am not sure, if this is what we want. From what I understand we are pruning all our testcases here. But this would result in a loss of past statement checks right? I would just stop generating a testcase here and return the current ones to keep the current coverage, rather than potentially losing it.

Priapos1004 replied:
Is irrelevant now

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -180,29 +184,138 @@ def _handle_compare_operation_cases(
+                        current_state = [{param_name: "a" * (target_len + 1)}]
+                    elif op == ">=":
+                        current_state = [{param_name: "a" * target_len}]
+                    elif op == "<":
```
Here and below. Couldn't we just generate a testcase with empty parameters here? It would still fit the criteria and give us more room for future length checks. 🤔

Priapos1004 replied:
Is irrelevant now

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -180,29 +184,138 @@ def _handle_compare_operation_cases(
+            # if neither exist
+            elif op in {"<", "<="}:
+                test_case[param1] = "aaa"
+                test_case[param2] = "aaaaa"
```
Again for more functionality I would use the shortest possible example like "a" and "" or "aa" and "a" but your version could provide more buffer for changes in future statements. 😄 I would still use "a" and "aaa" for the same difference in parameter length though 🤔

Lexolef replied:
Or "a" and "aaaa" considering the creation without current_state. Then it is even consistent 😁

Lexolef replied:
Is irrelevant now

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -211,17 +324,49 @@ def find_parameters(
 
                 value, param_name, is_single, param_left = self._get_compare_values(left, right)
+                # len-func logic
+                if op in {"==", "<", ">", ">=", "<="}:
```
What about "!=" you seem to check it in flip_op 🤔

Lexolef replied:
Is irrelevant now

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -211,17 +324,49 @@ def find_parameters(
+                # len-func logic
+                if op in {"==", "<", ">", ">=", "<="}:
+                    if (
+                        isinstance(left, Call)
```
I would summarize all these if statements similiar to _get_compare_values and simply create booleans for the Call checks of left and right first and then check these. This should also prevent ruff from having problems with two many checks here. Maybe you can even do a nested if, summarizing the last statement with one of the above for the call side. But I can see that this structure here seems more intuitive 😄

Lexolef replied:
Is irrelevant here

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -231,23 +376,44 @@ def find_parameters(
+        ):
+            method = node.func.attrname
+            expression = node.func.expr
+            param_name = ""
```
Why initialize param_name here? Isn't it only used within the next if statement? Shouldn't we just use param_name = expression.name as the first occurence of this parameter? 🤔

Lexolef replied:
Is irrelevant here

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -231,23 +376,44 @@ def find_parameters(
+                value, is_single = self._extract_literal_repr(arg)
+
+                if is_single:
+                    affix = value
```
This makes sense for readability. But for efficiency I would just call value directly in the method call below. But I guess it does not cost much 😅

Lexolef replied:
Is irrelevant now

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -277,10 +443,11 @@ def visit(
                     new_state = self.find_parameters(condition, old_state)
-                    current_state.extend(self.call_list_visit(node.body, new_state))
+                    temp_or_state.extend(self.call_list_visit(node.body, new_state))
+                current_state.extend(temp_or_state)
```
How is this necessary? Just for my overview. I am a bit confused here 😕

sharuiY replied:
In the earlier version the old_state is updated in each iteration, so that the next iteration use the updated old_state from the last iter. I thought this may cause problems, since we actually want to add each testcase for an or-statement independently.

Lexolef replied:
Stimmt! Aber irgendwie fixt es das bei mir nicht 😅 Ich schau mal ob es bei mir gerade liegt 🤔

Lexolef replied:
Okay ich habe auf meinem Branch jetzt einen Fix. Das Problem liegt darin, dass old_state in find_parameters übergeben wird und damit darin verändert, somit ist es beim zweiten Call noch das aus dem Letzten. Bei mir ist es jetzt: 
```
elif isinstance(node.test, BoolOp) and node.test.op == "or":
                old_state = copy.deepcopy(current_state)
                temp_or_state = []
                for condition in node.test.values:
                    old_state = copy.deepcopy(current_state)
                    new_state = self.find_parameters(condition, old_state)
                    temp_or_state.extend(self.call_list_visit(node.body, new_state))

                current_state = []
                current_state.extend(temp_or_state)
```
Ich würde meinen Branch auch noch nachher als PR hochladen. Deshalb kannst du den Part hier streichen oder so lassen, wie du willst. Ich kümmere mich dann nach deinem Merge um die Merge Conflicts 👍

Lexolef replied:
Oder du implementierst es noch. Das geht natürlich auch 😆 
Tschuldigung, dass ich übrigens zu Deutsch gewechselt habe. War gerade irgendwie nicht im richtigen Mindset 😓

sharuiY replied:
Stimmt, Ich habe dein Code hinzugefügt und probiert aber es gibt immer noch 2 Funktionen in mixed_with_len_func, die nicht abgedeckt werden können...

Lexolef replied:
Is irrelevant now

Lexolef commented on pyproject.toml:
```diff
@@ -6,3 +6,17 @@ exclude = [
+  "numpy >= 2.2.6",
+  "pydantic >= 2.11.7",
+  "pynguin",
+]
```
Not sure, if we should add/change this.

Lexolef replied:
Is irrelevant now

Lexolef commented on scripts/config.py:
```diff
@@ -6,7 +6,7 @@ class ExperimentSettings():
     """
     # Root directory of functions under test
-    EXAMPLES_DIR: str = "scripts/examples"
+    EXAMPLES_DIR: str = "scripts/examples/customized_examples_dir"
```
Please revert this before merging. I don't have customized_examples_dir 😆

sharuiY replied:
Have revert both files, it was not intended

Lexolef replied:
Is irrelevant now

Lexolef reviewed (commented):
It looks nice and seems to add a lot of functionality. I made some comments already, that you should check out. I let it run against the current main and it seems to not worsen any coverage we already had, so it looks good.
One last thing I would see, is that ruff seems to still have problems with a few things in this branch. Maybe use `ruff check` or `ruff check --fix` as seen in Samuels README.md to check these out. 😄

sharuiY committed:
Revert pyproject.toml and scripts/config.py. Refactor method find_parameters in tree_traverse_strategy.

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -229,33 +249,20 @@ def _handle_len(left: NodeNG, right: NodeNG, op: str, current_state: list[dict[s
-                    elif op == ">":
-                        current_state = [{param_name: "a" * (target_len + 1)}]
-                    elif op == ">=":
-                        current_state = [{param_name: "a" * target_len}]
```
We should keep these lines. So the > and >= statements.

Lexolef replied:
Is irrelevant here

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -264,32 +271,32 @@ def _handle_len_compare_between_params(
 
             # if only one exists
             elif len1 is not None:
                 new_len = len1 + 1 if op in {"<", "<="} else max(1, len1 - 1)
```
What is with == here? Do we just leave it out?

Lexolef replied:
Is irrelevant here

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -264,32 +271,32 @@ def _handle_len_compare_between_params(
-                test_case[param1] = "aaaaa"
-                test_case[param2] = "aaa"
+                test_case[param_l] = "a"
+                test_case[param_r] = ""
```
Same here. What about ==?

Lexolef replied:
Is irrelevant here

Lexolef commented on pynguin_0_41_0/src/pynguin/custom_seeding/strategy/tree_traverse_strategy.py:
```diff
@@ -445,7 +450,7 @@ def visit(self, node: NodeNG, current_state: list[dict[str, str]]) -> list[dict[
                 temp_or_state = []
                 for condition in node.test.values:
-                    new_state = self.find_parameters(condition, old_state)
+                    new_state = self.find_parameters(condition, copy.deepcopy(old_state))
```
This is smart, but I think we could just deepcopy current_state directly here now. Also you need to set current_state = [] before extending it below, since we don't want the statements before calling the or also generating testcases 😄

sharuiY committed:
Refactor _handle_len_compare_between_params & Add '==' checks in len nodes

Priapos1004 committed:
Refactored len-func and fixed several bugs

- little ruff changes to starts_with (TODO: refactore next)
- implemented not-logic changes from Levin's branch

Priapos1004 committed:
Added little changes to startswith, endswith logic

- added additional check in append_str_in to avoid unnecessary extension of testcase

Priapos1004 committed:
Ruff fix

Priapos1004 committed:
Merge branch 'main' into start_end_with_func

Priapos1004 committed:
Update tree_traverse_strategy.py

Priapos1004 committed:
Fixed little bugs
